galaxy_ng/app/api/ui/views/auth.py

```diff
@@ -55,6 +60,32 @@ def post(self, request, *args, **kwargs):
 class LogoutView(api_base.APIView):
     permission_classes = [access_policy.LogoutAccessPolicy]
 
+    def _oidc_logout(self, request):
+        logout_url_str = "{keycloak}/auth/realms/{realm}/protocol/openid-connect/logout"
+        logout_url = logout_url_str.format(keycloak=settings.KEYCLOAK_URL,
+                                           realm=settings.KEYCLOAK_REALM)
+
+        if not hasattr(request.user, 'social_auth'):
+            return
+
+        social = request.user.social_auth.get(provider="keycloak")
+        access_token = social.extra_data['access_token']
+        refresh_token = social.extra_data['refresh_token']
+        payload = {'client_id': settings.SOCIAL_AUTH_KEYCLOAK_KEY, 'refresh_token': refresh_token,
+                   'client_secret': settings.SOCIAL_AUTH_KEYCLOAK_SECRET}
+        headers = {"Authorization": "Bearer {access_token}".format(access_token=access_token)}
+        response = rpost(url=logout_url, headers=headers, data=payload)
+
+        if response.status_code == http_code.HTTP_200_OK:
+            log.debug("Logout of openid-connect client successful.")
+        else:
+            msg = "Log out for openid-connect failed with status {status} and content: {content}"
+            log.warning(msg.format(status=response.status_code, content=response.text))
+
     def post(self, request, *args, **kwargs):
+        # Trigger logout in Keycloak
+        if settings.get("SOCIAL_AUTH_KEYCLOAK_KEY"):
```

Comment:
If a user is logged in with django session auth, they can't log out.

Not a huge deal, but this could cause issues for people migrating from django to keycloak auth. Maybe run the regular logout of the the social auth logout fails?

```
api_1          | 172.18.0.10 - - [20/Aug/2021:18:40:24 +0000] "GET /api/automation-hub/_ui/v1/repo/published/?deprecated=false&offset=0&limit=10 HTTP/1.1" 200 3242 "http://localhost:8002/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0"
api_1          | Internal Server Error: /api/automation-hub/_ui/v1/auth/logout/
api_1          | Traceback (most recent call last):
api_1          |   File "/venv/lib64/python3.8/site-packages/django/core/handlers/exception.py", line 34, in inner
api_1          |     response = get_response(request)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/core/handlers/base.py", line 115, in _get_response
api_1          |     response = self.process_exception_by_middleware(e, request)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/core/handlers/base.py", line 113, in _get_response
api_1          |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
api_1          |     return view_func(*args, **kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/views/generic/base.py", line 71, in view
api_1          |     return self.dispatch(request, *args, **kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 509, in dispatch
api_1          |     response = self.handle_exception(exc)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 469, in handle_exception
api_1          |     self.raise_uncaught_exception(exc)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
api_1          |     raise exc
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 506, in dispatch
api_1          |     response = handler(request, *args, **kwargs)
api_1          |   File "/src/galaxy_ng/galaxy_ng/app/api/ui/views/auth.py", line 88, in post
api_1          |     self._oidc_logout(request)
api_1          |   File "/src/galaxy_ng/galaxy_ng/app/api/ui/views/auth.py", line 71, in _oidc_logout
api_1          |     social = request.user.social_auth.get(provider="keycloak")
api_1          |   File "/venv/lib64/python3.8/site-packages/django/db/models/manager.py", line 82, in manager_method
api_1          |     return getattr(self.get_queryset(), name)(*args, **kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/db/models/query.py", line 406, in get
api_1          |     raise self.model.DoesNotExist(
api_1          | social_django.models.UserSocialAuth.DoesNotExist: UserSocialAuth matching query does not exist.
api_1          | pulp [None]: django.request:ERROR: Internal Server Error: /api/automation-hub/_ui/v1/auth/logout/
api_1          | Traceback (most recent call last):
api_1          |   File "/venv/lib64/python3.8/site-packages/django/core/handlers/exception.py", line 34, in inner
api_1          |     response = get_response(request)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/core/handlers/base.py", line 115, in _get_response
api_1          |     response = self.process_exception_by_middleware(e, request)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/core/handlers/base.py", line 113, in _get_response
api_1          |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
api_1          |     return view_func(*args, **kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/views/generic/base.py", line 71, in view
api_1          |     return self.dispatch(request, *args, **kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 509, in dispatch
api_1          |     response = self.handle_exception(exc)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 469, in handle_exception
api_1          |     self.raise_uncaught_exception(exc)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
api_1          |     raise exc
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 506, in dispatch
api_1          |     response = handler(request, *args, **kwargs)
api_1          |   File "/src/galaxy_ng/galaxy_ng/app/api/ui/views/auth.py", line 88, in post
api_1          |     self._oidc_logout(request)
api_1          |   File "/src/galaxy_ng/galaxy_ng/app/api/ui/views/auth.py", line 71, in _oidc_logout
api_1          |     social = request.user.social_auth.get(provider="keycloak")
api_1          |   File "/venv/lib64/python3.8/site-packages/django/db/models/manager.py", line 82, in manager_method
api_1          |     return getattr(self.get_queryset(), name)(*args, **kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/db/models/query.py", line 406, in get
api_1          |     raise self.model.DoesNotExist(
api_1          | social_django.models.UserSocialAuth.DoesNotExist: UserSocialAuth matching query does not exist.
```